plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiTableFunction.java

```diff
@@ -0,0 +1,159 @@
+                        ScalarArgumentSpecification.builder()
+                                .name("QUERY")
+                                .type(VarcharType.VARCHAR)
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("START")
+                                .type(TIMESTAMP_TZ_NANOS)
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("END")
+                                .type(TIMESTAMP_TZ_NANOS)
+                                .build()),
+                GENERIC_TABLE);
+
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments, ConnectorAccessControl accessControl)
+    {
+        final ScalarArgument argument = (ScalarArgument) arguments.get("QUERY");
+        String query = ((Slice) argument.getValue()).toStringUtf8();
+
+        final LongTimestampWithTimeZone startArgument = (LongTimestampWithTimeZone) ((ScalarArgument) arguments.get("START")).getValue();
+        final LongTimestampWithTimeZone endArgument = (LongTimestampWithTimeZone) ((ScalarArgument) arguments.get("END")).getValue();
+
+        if (Strings.isNullOrEmpty(query)) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, query);
+        }
+        requireNonNull(startArgument);
+        requireNonNull(endArgument);
+        requireNonNull(metadata);
+
+        // determine the returned row type
+        List<ColumnHandle> columnHandles;
+        try {
+            columnHandles = metadata.getColumnHandles(query);
+        }
+        catch (UndeclaredThrowableException e) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Cannot get column definition", Throwables.getRootCause(e));
+        }
+
+        checkArgument(!columnHandles.isEmpty(), "Cannot get column definition");
+
+        Descriptor returnedType = new Descriptor(columnHandles.stream()
+                .map(LokiColumnHandle.class::cast)
+                .map(column -> new Descriptor.Field(column.name(), Optional.of(column.type())))
+                .collect(toImmutableList()));
+
+        Instant end = toInstant(endArgument);
+        Instant start = toInstant(startArgument);
+
+        final LokiTableHandle tableHandle = new LokiTableHandle(
+                query,
+                start,
+                end,
+                columnHandles);
+
+        return TableFunctionAnalysis.builder()
+                .returnedType(returnedType)
+                .handle(new QueryHandle(tableHandle))
+                .build();
+    }
+
+    /**
+     * Convert LongTimestampWithTimeZone to Instant. See BigQuery code for an example
+     */
+    private Instant toInstant(LongTimestampWithTimeZone timestamp)
```

Comment:
But `Instant` are not Zoned aware right ? Can we use `ZonedDateTime` which would use Zone which is specified as a part of the user criteria -

Reply:
> But Instant are not Zoned aware right ?

They are not but they are always UTC. We are basically converting the timestamp with a time zone to nano seconds since 1970-01-01T00:00:00Z.